File: behaviour-tree-test/js/nodes/sequence.js
/**
 * A SEQUENCE node.
 * The child nodes are executed in sequence until one fails or all succeed.
 * @param children The child nodes. 
 */
function Sequence(children) {
    /**
     * The node state.
     */
    let state = NodeState.READY;
   
    /**
     * Update the node and get whether the node state has changed.
     * @returns Whether the state of this node has changed as part of the update.
     */
    this.update = function() {
        // Get the pre-update node state.
        const initialState = state;

        // Iterate over all of the children of this node.
        for (const child of children) {
            // If the child has never been updated then we will need to do this now.
            if (child.getState() === NodeState.READY) {
                child.update();
            }

            // If the current child has a state of 'SUCCEEDED' then we should move on to the next child.
            if (child.getState() === NodeState.SUCCEEDED) {
                // Find out if the current child is the last one in the sequence.
                // If it is then this sequence node has also succeeded.
                if (children.indexOf(child) === children.length - 1) {
                    // This node is a 'SUCCEEDED' node.
                    state = NodeState.SUCCEEDED;

                    // There is no need to check the rest of the sequence as we have completed it.
                    return state !== initialState;
                } else {
                    // The child node succeeded, but we have not finished the sequence yet.
                    continue;
                }
            }

            // If the current child has a state of 'FAILED' then this node is also a 'FAILED' node.
            if (child.getState() === NodeState.FAILED) {
                // This node is a 'FAILED' node.
                state = NodeState.FAILED;

                // There is no need to check the rest of the sequence.
                return state !== initialState;
            }

            // The node should be in the 'RUNNING' state.
            if (child.getState() === NodeState.RUNNING) {
                // This node is a 'RUNNING' node.
                state = NodeState.RUNNING;

                // There is no need to check the rest of the sequence as the current child is still running.
                return state !== initialState;
            }

            // The child node was not in an expected state.
            throw "Error: child node was not in an expected state.";
        }
    };

    /**
     * Gets the state of the node.
     */
    this.getState = () => state;

    /**
     * Reset the state of the node.
     */
    this.reset = () => {
        // Reset the state of this node.
        state = NodeState.READY;

        // Reset each child node.
        children.forEach((child) => child.reset());
    };
};

File: Dungeony/core/src/com/dumbpug/mistreevous/node/Sequence.java
package com.dumbpug.mistreevous.node;

import com.dumbpug.mistreevous.IBoard;
import com.dumbpug.mistreevous.State;
import com.dumbpug.mistreevous.decorator.Decorators;
import java.util.ArrayList;

/**
 * A SEQUENCE composite node.
 * The child nodes are executed in sequence until one fails or all succeed.
 */
public class Sequence extends Composite {
    /**
     * Creates a new instance of the Sequence class.
     * @param decorators The node decorators.
     * @param children The child node of this composite node.
     */
    public Sequence(Decorators decorators, ArrayList<Node> children) {
        super(decorators, children);
    }

    @Override
    public void onUpdate(IBoard board) {
        // Iterate over all of the children of this node.
        for (Node child : this.getChildren()) {
            // If the child has never been updated or is running then we will need to update it now.
            if (child.is(State.READY) || child.is(State.RUNNING)) {
                // Update the child of this node.
                child.update(board);
            }

            // If the current child has a state of 'SUCCEEDED' then we should move on to the next child.
            if (child.is(State.SUCCEEDED)) {
                // Find out if the current child is the last one in the sequence.
                // If it is then this sequence node has also succeeded.
                if (this.getChildren().indexOf(child) == this.getChildren().size() - 1) {
                    // This node is a 'SUCCEEDED' node.
                    this.setState(State.SUCCEEDED);

                    // There is no need to check the rest of the sequence as we have completed it.
                    return;
                } else {
                    // The child node succeeded, but we have not finished the sequence yet.
                    continue;
                }
            }

            // If the current child has a state of 'FAILED' then this node is also a 'FAILED' node.
            if (child.is(State.FAILED)) {
                // This node is a 'FAILED' node.
                this.setState(State.FAILED);

                // There is no need to check the rest of the sequence.
                return;
            }

            // The node should be in the 'RUNNING' state.
            if (child.is(State.RUNNING)) {
                // This node is a 'RUNNING' node.
                this.setState(State.RUNNING);

                // There is no need to check the rest of the sequence as the current child is still running.
                return;
            }

            // The child node was not in an expected state.
            throw new RuntimeException("A child node of a SEQUENCE node was not in an expected state.");
        }
    }

    @Override
    public void validate() {
        // A sequence node must have at least a single child.
        if (this.getChildren().isEmpty()) {
            throw new RuntimeException("A SEQUENCE node must have at least a single child node.");
        }
    }

    @Override
    public NodeType getNodeType() {
        return NodeType.SEQUENCE;
    }
}
